From 3caa3c773fee68a82b6349365cf142f3fa83f0f7 Mon Sep 17 00:00:00 2001
From: Srinivas Pandruvada <srinivas.pandruvada@linux.intel.com>
Date: Mon, 14 Mar 2022 13:06:57 -0700
Subject: [PATCH 3/3] thermal: int340x: Add Resume callback

Add resume callback to issue _OSC for last stored policy. It is possible
that during suspend/hibernate, firmware didn't store the settings. So
restore during resume.

Signed-off-by: Srinivas Pandruvada <srinivas.pandruvada@linux.intel.com>
---
 .../intel/int340x_thermal/int3400_thermal.c   | 41 ++++++++++++++++++-
 1 file changed, 39 insertions(+), 2 deletions(-)

diff --git a/drivers/thermal/intel/int340x_thermal/int3400_thermal.c b/drivers/thermal/intel/int340x_thermal/int3400_thermal.c
index b06a70fc6c18..5d0ed7fa9dc4 100644
--- a/drivers/thermal/intel/int340x_thermal/int3400_thermal.c
+++ b/drivers/thermal/intel/int340x_thermal/int3400_thermal.c
@@ -640,6 +640,42 @@ static int int3400_thermal_remove(struct platform_device *pdev)
 	return 0;
 }
 
+
+#ifdef CONFIG_PM_SLEEP
+static int int3400_thermal_resume(struct device *dev)
+{
+	struct int3400_thermal_priv *priv = dev_get_drvdata(dev);
+
+	if (!priv)
+		return 0;
+
+	if (priv->current_uuid_index >= 0 &&
+	    priv->current_uuid_index < INT3400_THERMAL_MAXIMUM_UUID) {
+		int enabled = 1;
+
+		int3400_thermal_run_osc(priv->adev->handle,
+					int3400_thermal_uuids[priv->current_uuid_index],
+					&enabled);
+	}
+
+	if (priv->os_uuid_mask) {
+		int cap;
+
+		cap = ((priv->os_uuid_mask << 1) | 0x01);
+		int3400_thermal_run_osc(priv->adev->handle,
+					"b23ba85d-c8b7-3542-88de-8de2ffcfd698", &cap);
+	}
+
+	evaluate_odvp(priv);
+
+	return 0;
+}
+#else
+#define int3400_thermal_resume NULL
+#endif
+
+static SIMPLE_DEV_PM_OPS(int3400_thermal_pm, NULL, int3400_thermal_resume);
+
 static const struct acpi_device_id int3400_thermal_match[] = {
 	{"INT3400", 0},
 	{"INTC1040", 0},
@@ -654,8 +690,9 @@ static struct platform_driver int3400_thermal_driver = {
 	.probe = int3400_thermal_probe,
 	.remove = int3400_thermal_remove,
 	.driver = {
-		   .name = "int3400 thermal",
-		   .acpi_match_table = ACPI_PTR(int3400_thermal_match),
+			.name = "int3400 thermal",
+			.acpi_match_table = ACPI_PTR(int3400_thermal_match),
+			.pm = &int3400_thermal_pm,
 		   },
 };
 
-- 
2.35.1

